src/uartproxy.c

```diff
@@ -193,7 +212,7 @@ int uartproxy_run(struct uartproxy_msg_start *start)
                     break;
                 exc_count = 0;
                 exc_guard = GUARD_RETURN;
-                checksum_val = checksum((void *)request.mrequest.addr, request.mrequest.size);
+                checksum_val = data_checksum((void *)request.mrequest.addr, request.mrequest.size);
```

Comment:
Ahh, but this won't do what I want. The goal of the sentinel is to catch data offset errors (e.g. bytes were added or removed), but that only works if the checksum comes *after* the data, which is not the case for data transfers, only for the commands and events.

Can you *add* a new sentinel field after the data to the data transfer ops? This does increase overhead slightly, but it is unlikely to matter given transfer overhead. With the new iodev_queue thing, this can be done without triggering unnecessary USB packet splitting overhead. You can use a different sentinel for this from the one used for checksum_val, so that if some weird alignment happens it won't falsely pass.

Reply:
I added the extra sentinel for both reads and writes. I’m not completely sure if [the way I flush queued writes](https://github.com/AsahiLinux/m1n1/commit/653a5497bbf239a9534ccb200fde72d22394f68d#diff-f84ef3ef6c6fcb6f5e938547f2a8f6cd000e68c6fd7a39639a40e19fbdd7362bR283) in the reply handler is correct, though.